Prefill declaration sections

TheDoubleK opened the pull request:
Here we have working code that pretty much completes:
https://www.pivotaltracker.com/story/show/138702601 (Pre-filling relevant declaration sections with answers from a previous framework)
https://www.pivotaltracker.com/story/show/138703089 (Adding blue banners and validation-style blue messages to pre-filled pages)

~~**NOTE**~~ 
~~!!!!!!!!! There are no new unit tests at all for this new behaviour :scream:  !!!!!!!!!~~

~~Feel free to pick this up and add some relevant tests.~~

~~Or review it and merge it in if you feel like seeing it all working on Friday, and on my honour I will add the required unit tests on Monday.~~

**UPDATE 26/2: Tests added and rebased up-to-date - now 100% ready for review.**

Here's a screenshot:
![screen shot 2017-02-26 at 12 13 36](https://cloud.githubusercontent.com/assets/6525554/23339578/b6cd2644-fc1d-11e6-9131-be2e5d43e6dd.png)

risicle commented on tests/app/main/test_frameworks.py:
```diff
@@ -2805,14 +2810,188 @@ def test_get_with_with_previous_answers(self, data_api_client):
+                prefill_declaration_from_framework_slug="digital-outcomes-and-specialists-2"
+            )
+            # The previous declaration to prefill from
+            data_api_client.get_supplier_declaration.return_value = {
```
We don't assert the call args for `get_supplier_declaration` anywhere here do we?

I added some useful tools at the top of this file `_assert_args_and_raise` and `_assert_args_and_return` which assert the args & kwargs in-the-loop ( only useful for cases where mocked functions should get called only once per test).

benvand replied:
I'm going to suggest leaving these functions alone for now. They're in the wrong place at the moment.
Happy to discuss these methods' inclusion and location another time.

I'd also say we don't strictly need to assert the calls in this test:
This test asserts the existence of elements and their state on the page. What this test asserts can be achieved entirely independently of the api calls.

If you want, on the other hand, to assert that a given piece of code makes certain calls then check the calls but not the html.

A caveat to this might be if you wanted to ensure that the state on the page can be achieved only with a certain number of api calls, however that's not what _this_ test is meant to do.


I'm not _that_ fussed if we check the calls here or not but I'd question wether or not it's causing fragility in the test suite by enforcing conditions we actually don't care about.

risicle replied:
I would say we care about *all* these conditions *all* the time. "Separate tests for separate concerns" *sounds* nice, but whichever way you look at it, doesn't end up testing as much. Only if this exact function-under-test call was made with the exact same setup prologue would it be not worth making these extra assertions here. And if that were the case, then I would advocate merging those tests to improve run time.

This comes back to the coverage argument. We're kidding ourselves with any coverage reports we're doing if we're not actually asserting all that we could be in all cases.

e.g. from what I can tell, nowhere else in the tests do we test the case where the function under test is called with the declaration being `started`, not having any pre-existing answers and having the `prefill_declaration_from_framework_slug` set. We'd be missing out on the opportunity to ensure that combination doesn't go crazy and idunno call `get_supplier_declaration` for `g-cloud-9` in this case. And we do *care* that it doesn't do that as if it doesn't call it with the right argument, there's no way it can possibly give us the right answer under real life conditions...

benvand replied:
What you're describing above isn't unit testing. 
A test with more condition checks may cover more code but they are inherently more fragile and add far less value by being less indicative of an overall problem.

Put it this way, we could check, in every one of our unit tests, whether the GOV.UK cookie banner appears.
Say we then add a check to every test to make sure the CrownCopyright appears in the bottom right.

Then, on the next test run, everything breaks. 
Was it the banner or the CC or both?

We don't do the above because it doesn't add value, every test would just break if we changed the banner.
Unless you're checking one specific thing unit test failures help _narrow down_ a problem at best and at worst point you completely in the wrong direction.

Obviously this is a massive oversimplification but it demonstrates why it's important to check only what you want to find out. Tests are cheap, once you've written one it's very easy to write more so if needs be then split them out into several tests that functionally do the same thing but check different conditions.

Then when a test breaks called `test_view_makes_only_one_api_call` fails you know _exactly_ what's wrong instead of a test called `test_declaration_page` failing because someone updated the frontend toolkit.

TheDoubleK replied:
It feels to me like testing _all_ the API calls made by a view could fall into the category of "I know too much about the inside of the function under test" kind of testing.

risicle replied:
I don't tend to get too bound up by terms, what's "unit testing" or not "unit testing" - I'm just concerned with what's useful testing (hey maybe I should register _that_ term). When a test fails for me, the first thing I look at isn't the test name (take a survey of test names we have - more often that not they aren't particularly illuminating), I run it under `--pdb` and look at the assertion that failed, which hopefully is written in a way that makes it clear what it's trying to assert for and why.

Yes, testing for the govuk banner is going too far, but this is where you have to apply some pragma. It's also something that should have been tested as part of its own module. Code such as a view should be able to consider something as fundamental as the banner to be a black box that is tested on its own and "works".

And while "tests are cheap", maintaining them certainly is not. How may people when they fix up a particular small test go and similarly fix up the 57 other variants of the test sitting next to it? Copy pasted test variants have a tendency to diverge... 

Test execution time is also something that isn't particularly cheap.

@TheDoubleK I think only when the situation falls into the category of "leave the function some leeway of how it wants to achieve this, but if it's "there's no possible way for it to give the right answer if it doesn't call the right thing", I'd say not. If we want to go mad, I'd say the right thing to do would be for the `get_supplier_declaration` mock to be able to return appropriate mock declarations for either framework depending on how it's called and for the assertions to be able to distinguish between them. But that would be insane to set up manually all the time, which is one reason I'd like to move to a more "comprehensive" api-mocking system.

risicle commented on tests/app/main/test_frameworks.py:
```diff
@@ -2805,14 +2810,188 @@ def test_get_with_with_previous_answers(self, data_api_client):
+            assert len(doc.xpath('//input[@id="input-organisedCrime-no"]/@checked')) == 1
+
+            # Blue banner message is shown at top of page
+            assert doc.xpath('normalize-space(//div[@class="banner-information-without-action"]/p/text())') == \
```
`normalize-space(string(//div[@class="banner-information-without-action"]))` would be more robust here but meh

risicle commented on tests/app/main/test_frameworks.py:
```diff
@@ -2805,14 +2810,188 @@ def test_get_with_with_previous_answers(self, data_api_client):
+            info_messages = doc.xpath('//div[@class="message-wrapper"]//span[@class="message-content"]')
+            assert len(info_messages) == 0
+
+    def test_answers_not_prefilled_if_section_marked_as_prefill_false(self, data_api_client):
```
Do we test a case where we're trying to prefill from a framework that is missing some of the desired questions?

TheDoubleK replied:
No but we can do.

benvand commented on app/templates/frameworks/edit_declaration_section.html:
```diff
@@ -52,11 +59,19 @@
+                    message="This answer is from your {} declaration".format(section_has_been_prefilled_from)
+                            if section_has_been_prefilled_from and
+                                (declaration_answers.get(question.id) is not none or question.type == 'multiquestion')
+                            else None
```
It would probably look even cleaner without the ternary tbf, especially in Jinja syntax

risicle replied:
/me would probably do

```
message=(
        "This answer is from your {} declaration".format(section_has_been_prefilled_from)
    if
        section_has_been_prefilled_from and (declaration_answers.get(question.id) is not none or question.type == 'multiquestion')
    else
        None
)
```

though probably not to everyone's tastes.

TheDoubleK replied:
Not sure how to avoid the ternary without duplicating the whole thing in an `if...else` block (as it used to be), so going to go with @risicle's suggestion which makes it at least a bit easier to read I think.

risicle commented:
(hold on maybe I should actually try running it before I pass judgement...)

benvand commented on tests/app/helpers.py:
```diff
@@ -309,7 +309,7 @@ def framework(
                 ]
             metaframework = "g-cloud"
-        elif slug == 'digital-outcomes-and-specialists':
+        elif 'digital-outcomes-and-specialists' in slug:
```
Fo sho

TheDoubleK replied:
It's all fine, nothing to see here :trollface:

TheDoubleK commented on app/templates/macros/toolkit_forms.html:
```diff
@@ -1,9 +1,9 @@
-{% macro text(question_content, service_data, errors, question_number=None, get_question=None) -%}
```
The problem is that because the macros are called dynamically they have to all accept the same arguments, so without using `kwargs` in order to add a new argument to be used in one macro we have to add it to the argument list of all macros.  So this seems the exact kind of thing that `kwargs` is designed for?

TheDoubleK replied:
What is the standard python way of doing things?  And if we're in a Jinja template does it still apply?  I only just found out how kwargs works in Jinja so may well be abusing it here through not knowing any better.  But it seemed like a neater solution than adding unused arguments to all macros.

risicle replied:
Yeeeeeeah.... in python, a function defined with only positional parameters would only accept those parameters, you would have to explicitly add a `**kwargs` parameter to allow that, so the syntax diverges a bit here - still, if there's a reason for it, go right ahead. I was just making a comment in case you were doing it just for the hell of it.

TheDoubleK replied:
Yep this is just how Jinja does kwargs I'm afraid - you can't pass them in explicitly.  Pretty horrible design choice, I agree :/

TheDoubleK commented on app/main/views/frameworks.py:
```diff
@@ -471,21 +471,42 @@ def framework_supplier_declaration(framework_slug, section_id=None):
+                    prefill_from_slug
+                )['declaration']
+                all_answers = declaration_to_reuse
+                section_has_been_prefilled_from = framework_to_reuse['name']
```
Yep - will fix.

TheDoubleK replied:
I'm going with `name_of_framework_that_section_has_been_prefilled_from` - or is that way too long?

TheDoubleK commented on tests/app/main/test_frameworks.py:
```diff
@@ -2805,14 +2810,188 @@ def test_get_with_with_previous_answers(self, data_api_client):
         with self.app.test_client():
             self.login()
+            # First call is for the current framework; second call for the framework to pre-fill from
+            data_api_client.get_framework.side_effect = [
```
Fair point.  Fixing up.

TheDoubleK commented:
Comments addressed:
* Changed variable name to `name_of_framework_that_section_has_been_prefilled_from` to make it clearer that it is a framework name being passed round and not some other sort of object.
* Fixed tests to not depend on order of calls to `data_api_client.get_framework`
* Add test for case where previous declaration doesn't have answers for all questions in a section so page is only partially pre-filled.
* Use more robust xPath expression for retrieving banner message
* Assert call args to `data_api_client.get_supplier_declaration` in cases where it should be called.

risicle commented on app/templates/frameworks/edit_declaration_section.html:
```diff
@@ -52,11 +59,22 @@
+                    message=(
+                            "This answer is from your {} declaration".format(name_of_framework_that_section_has_been_prefilled_from)
+                        if
+                            name_of_framework_that_section_has_been_prefilled_from and (declaration_answers.get(question.id) is not none or question.type == 'multiquestion')
```
Lowercase `none` here - mistake or jinja thing?

galund replied:
I think this highlights a problem - these macros are basically causing us to write Python code in our templates. The benefit from separation of concerns that templating is supposed to bring is not being realised right now, is the point. These definitely aren't things that could be owned by front-end designers/specialists, in their current state, IMO.

risicle replied:
Additionally, `or question.type == 'multiquestion'` here seems to cause the "this answer is from your..." message to appear even on questions that have a missing previous entry as long as they're multiquestions.
![20170227missingprefillmultiquestions](https://cloud.githubusercontent.com/assets/807447/23370238/008f1078-fd0c-11e6-95ce-40c4fa2aa80c.png)

risicle replied:
(though maybe this isn't important if we're only catering for all-prefilled-or-none-prefilled cases)

TheDoubleK replied:
Lowercase none is a jinja thing.

Yep the multiquestion thing will need an extra condition in there, good spot, will fix. (At the moment it is all or nothing though.)

George, philosophical questions aside that's a much bigger issue than this pull request can address right now.

risicle commented:
Anyway, given the changes, I'm sure this will do for G9. :+1:

risicle commented:
Before you merge - should this PR also take care of removing the ability for the "final page" to mark the declaration as `complete`?
